src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/AutoInsert/AutoInsertService.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.AutoInsert;
+
+internal class AutoInsertService(IEnumerable<IOnAutoInsertProvider> onAutoInsertProviders) : IAutoInsertService
+{
+    private readonly ImmutableArray<IOnAutoInsertProvider> _onAutoInsertProviders = onAutoInsertProviders.ToImmutableArray();
+
+    public static FrozenSet<string> HtmlAllowedAutoInsertTriggerCharacters { get; }
+        = new string[] { "=" }.ToFrozenSet(StringComparer.Ordinal);
+    public static FrozenSet<string> CSharpAllowedAutoInsertTriggerCharacters { get; }
+        = new string[] { "'", "/", "\n" }.ToFrozenSet(StringComparer.Ordinal);
+
+    private readonly ImmutableArray<string> _triggerCharacters = CalculateTriggerCharacters(onAutoInsertProviders);
+
+    private static ImmutableArray<string> CalculateTriggerCharacters(IEnumerable<IOnAutoInsertProvider> onAutoInsertProviders)
+    {
+        var builder = ImmutableArray.CreateBuilder<string>(onAutoInsertProviders.Count());
+        foreach (var provider in onAutoInsertProviders)
+        {
+            var triggerCharacter = provider.TriggerCharacter;
+            if (!builder.Contains(triggerCharacter))
```

Comment:
I understand and was aware of O(n^2), but considering we have only 2 providers and don't anticipate more, seems like the overhead of using a HashSet here would be more than overhead of O(n^2). I'll add the HashSet if you feel strongly about it, otherwise I'll just stick with .Any (which is also O(n^2)) only because of the extremely small n, which is 2, and is unlikely to change, at least not in any significant way.